File: src/feishu_auth_kit/orchestration.py
from __future__ import annotations

import json
import os
import uuid
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any, Literal

import requests

from .models import DeviceAuthorization
from .runtime_cards import (
    ContinuationState,
    FileContinuationStore,
    RuntimeCard,
    build_device_flow_card,
    build_permission_missing_card,
)
from .scopes import batch_scopes, filter_sensitive_scopes

AuthErrorKind = Literal["app_scope_missing", "user_auth_required", "user_scope_insufficient"]
AuthDecision = Literal["permission_card", "device_flow", "authorized"]
ScopeNeedType = Literal["one", "all"]
TokenType = Literal["tenant", "user"]


def _dedupe_preserve_order(items: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for item in items:
        normalized = item.strip()
        if normalized and normalized not in seen:
            seen.add(normalized)
            result.append(normalized)
    return result


def _default_pending_flow_store_path() -> Path:
    configured = os.getenv("FEISHU_AUTH_KIT_PENDING_FLOW_STORE")
    if configured:
        return Path(configured).expanduser()
    state_home = os.getenv("XDG_STATE_HOME")
    if state_home:
        base = Path(state_home).expanduser()
    else:
        base = Path.home() / ".local" / "state"
    return base / "feishu-auth-kit" / "pending_flows.json"


@dataclass(frozen=True)
class PendingAuthFlow:
    flow_key: str
    operation_id: str
    app_id: str
    user_open_id: str | None
    decision: AuthDecision
    required_scopes: list[str] = field(default_factory=list)
    token_type: TokenType = "user"
    scope_need_type: ScopeNeedType = "all"
    status: str = "pending"
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class PendingFlowUpsertResult:
    flow: PendingAuthFlow
    reused: bool
    merged_scopes: list[str] = field(default_factory=list)


class FilePendingFlowRegistry:
    def __init__(self, path: str | Path | None = None) -> None:
        self.path = Path(path).expanduser() if path else _default_pending_flow_store_path()

    def _read_all(self) -> dict[str, dict[str, Any]]:
        if not self.path.exists():
            return {}
        payload = json.loads(self.path.read_text(encoding="utf-8"))
        if not isinstance(payload, dict):
            return {}
        items = payload.get("pending_flows", payload)
        if not isinstance(items, dict):
            return {}
        return {str(key): value for key, value in items.items() if isinstance(value, dict)}

    def _write_all(self, items: dict[str, dict[str, Any]]) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        temp_path = self.path.with_suffix(f"{self.path.suffix}.tmp")
        temp_path.write_text(
            json.dumps({"pending_flows": items}, indent=2, sort_keys=True) + "\n",
            encoding="utf-8",
        )
        temp_path.replace(self.path)

    def load(self, flow_key: str) -> PendingAuthFlow | None:
        item = self._read_all().get(flow_key)
        if not item:
            return None
        return PendingAuthFlow(
            flow_key=str(item["flow_key"]),
            operation_id=str(item["operation_id"]),
            app_id=str(item["app_id"]),
            user_open_id=item.get("user_open_id"),
            decision=item["decision"],
            required_scopes=list(item.get("required_scopes") or []),
            token_type=item.get("token_type", "user"),
            scope_need_type=item.get("scope_need_type", "all"),
            status=str(item.get("status", "pending")),
            metadata=item.get("metadata") or {},
        )

    def remove(self, flow_key: str) -> bool:
        items = self._read_all()
        removed = items.pop(flow_key, None)
        if removed is None:
            return False
        self._write_all(items)
        return True

    def upsert(self, flow: PendingAuthFlow) -> PendingFlowUpsertResult:
        items = self._read_all()
        existing = self.load(flow.flow_key)
        if existing is None:
            items[flow.flow_key] = asdict(flow)
            self._write_all(items)
            return PendingFlowUpsertResult(
                flow=flow,
                reused=False,
                merged_scopes=list(flow.required_scopes),
            )

        merged_scopes = _dedupe_preserve_order(
            [*existing.required_scopes, *flow.required_scopes]
        )
        merged_flow = PendingAuthFlow(
            flow_key=existing.flow_key,
            operation_id=existing.operation_id,
            app_id=existing.app_id,
            user_open_id=flow.user_open_id or existing.user_open_id,
            decision=flow.decision,
            required_scopes=merged_scopes,
            token_type=flow.token_type,
            scope_need_type=flow.scope_need_type,
            status=flow.status,
            metadata={**existing.metadata, **flow.metadata},
        )
        items[flow.flow_key] = asdict(merged_flow)
        self._write_all(items)
        return PendingFlowUpsertResult(
            flow=merged_flow,
            reused=True,
            merged_scopes=merged_scopes,
        )


@dataclass(frozen=True)
class ScopeAuthorizationPlan:
    requested_scopes: list[str]
    app_granted_scopes: list[str]
    user_granted_scopes: list[str]
    already_granted_scopes: list[str]
    missing_user_scopes: list[str]
    unavailable_scopes: list[str]
    batches: list[list[str]]


def plan_scope_authorization(
    *,
    requested_scopes: list[str],
    app_granted_scopes: list[str],
    user_granted_scopes: list[str],
    batch_size: int = 100,
    filter_sensitive: bool = True,
) -> ScopeAuthorizationPlan:
    requested = _dedupe_preserve_order(requested_scopes)
    app_granted = _dedupe_preserve_order(app_granted_scopes)
    user_granted = _dedupe_preserve_order(user_granted_scopes)
    if filter_sensitive:
        requested = filter_sensitive_scopes(requested)
        app_granted = filter_sensitive_scopes(app_granted)
        user_granted = filter_sensitive_scopes(user_granted)
    app_granted_set = set(app_granted)
    user_granted_set = set(user_granted)
    unavailable = [scope for scope in requested if scope not in app_granted_set]
    available = [scope for scope in requested if scope in app_granted_set]
    already = [scope for scope in available if scope in user_granted_set]
    missing = [scope for scope in available if scope not in user_granted_set]
    return ScopeAuthorizationPlan(
        requested_scopes=requested,
        app_granted_scopes=app_granted,
        user_granted_scopes=user_granted,
        already_granted_scopes=already,
        missing_user_scopes=missing,
        unavailable_scopes=unavailable,
        batches=batch_scopes(missing, batch_size=batch_size),
    )


@dataclass(frozen=True)
class AuthRequirement:
    error_kind: AuthErrorKind
    required_scopes: list[str]
    token_type: TokenType = "user"
    scope_need_type: ScopeNeedType = "all"
    user_open_id: str | None = None
    flow_key: str | None = None
    operation_id: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class AuthContinuation:
    operation_id: str
    flow_key: str
    app_id: str
    decision: AuthDecision
    user_open_id: str | None
    required_scopes: list[str]
    token_type: TokenType
    scope_need_type: ScopeNeedType
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_state(self) -> ContinuationState:
        return ContinuationState(
            operation_id=self.operation_id,
            app_id=self.app_id,
            kind=self.decision,
            status="waiting",
            payload={
                "flow_key": self.flow_key,
                "user_open_id": self.user_open_id,
                "required_scopes": self.required_scopes,
                "token_type": self.token_type,
                "scope_need_type": self.scope_need_type,
                "metadata": self.metadata,
            },
        )

    @classmethod
    def from_state(cls, state: ContinuationState) -> AuthContinuation:
        payload = state.payload
        return cls(
            operation_id=state.operation_id,
            flow_key=str(payload["flow_key"]),
            app_id=state.app_id,
            decision=state.kind,  # type: ignore[arg-type]
            user_open_id=payload.get("user_open_id"),
            required_scopes=list(payload.get("required_scopes") or []),
            token_type=payload.get("token_type", "user"),
            scope_need_type=payload.get("scope_need_type", "all"),
            metadata=payload.get("metadata") or {},
        )


def save_auth_continuation(
    store: FileContinuationStore,
    continuation: AuthContinuation,
) -> ContinuationState:
    state = continuation.to_state()
    store.save(state)
    return state


def load_auth_continuation(
    store: FileContinuationStore,
    operation_id: str,
) -> AuthContinuation | None:
    state = store.load(operation_id)
    if state is None:
        return None
    return AuthContinuation.from_state(state)


@dataclass(frozen=True)
class RoutedAuthAction:
    decision: AuthDecision
    flow: PendingAuthFlow
    reused_existing_flow: bool
    continuation: AuthContinuation
    card: RuntimeCard | None = None
    unavailable_scopes: list[str] = field(default_factory=list)
    batches: list[list[str]] = field(default_factory=list)


def route_auth_requirement(
    *,
    app_id: str,
    requirement: AuthRequirement,
    pending_flows: FilePendingFlowRegistry,
    continuation_store: FileContinuationStore,
    permission_url: str | None = None,
    authorization: DeviceAuthorization | None = None,
) -> RoutedAuthAction:
    decision: AuthDecision
    if requirement.error_kind == "app_scope_missing":
        decision = "permission_card"
    else:
        decision = "device_flow"

    flow_key = (
        requirement.flow_key
        or f"{app_id}:{requirement.user_open_id or 'unknown'}:{decision}"
    )
    operation_id = requirement.operation_id or uuid.uuid4().hex
    upsert = pending_flows.upsert(
        PendingAuthFlow(
            flow_key=flow_key,
            operation_id=operation_id,
            app_id=app_id,
            user_open_id=requirement.user_open_id,
            decision=decision,
            required_scopes=list(requirement.required_scopes),
            token_type=requirement.token_type,
            scope_need_type=requirement.scope_need_type,
            metadata=requirement.metadata,
        )
    )
    continuation = AuthContinuation(
        operation_id=upsert.flow.operation_id,
        flow_key=upsert.flow.flow_key,
        app_id=app_id,
        decision=decision,
        user_open_id=upsert.flow.user_open_id,
        required_scopes=upsert.merged_scopes,
        token_type=upsert.flow.token_type,
        scope_need_type=upsert.flow.scope_need_type,
        metadata=upsert.flow.metadata,
    )
    save_auth_continuation(continuation_store, continuation)

    if decision == "permission_card":
        if not permission_url:
            raise ValueError("permission_url is required for app_scope_missing routing")
        card = build_permission_missing_card(
            app_id=app_id,
            operation_id=upsert.flow.operation_id,
            missing_scopes=upsert.merged_scopes,
            permission_url=permission_url,
            user_open_id=upsert.flow.user_open_id,
        )
        return RoutedAuthAction(
            decision=decision,
            flow=upsert.flow,
            reused_existing_flow=upsert.reused,
            continuation=continuation,
            card=card,
        )

    if authorization is None:
        raise ValueError("authorization is required for user auth routing")
    card = build_device_flow_card(
        app_id=app_id,
        operation_id=upsert.flow.operation_id,
        authorization=authorization,
    )
    return RoutedAuthAction(
        decision=decision,
        flow=upsert.flow,
        reused_existing_flow=upsert.reused,
        continuation=continuation,
        card=card,
    )


@dataclass(frozen=True)
class SyntheticRetryArtifact:
    schema: str
    kind: str
    operation_id: str
    app_id: str
    user_open_id: str | None
    text: str
    reason: str
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def build_synthetic_retry_artifact(
    *,
    operation_id: str,
    app_id: str,
    user_open_id: str | None,
    text: str,
    reason: str,
    metadata: dict[str, Any] | None = None,
) -> SyntheticRetryArtifact:
    return SyntheticRetryArtifact(
        schema="feishu-auth-kit.synthetic-retry.v1",
        kind="synthetic_retry",
        operation_id=operation_id,
        app_id=app_id,
        user_open_id=user_open_id,
        text=text,
        reason=reason,
        metadata=metadata or {},
    )


@dataclass(frozen=True)
class IdentityVerificationResult:
    valid: bool
    expected_open_id: str
    actual_open_id: str | None = None


def verify_access_token_identity(
    *,
    access_token: str,
    expected_open_id: str,
    brand: str = "feishu",
    session: Any | None = None,
    timeout: int = 30,
) -> IdentityVerificationResult:
    from .domains import resolve_domains

    current_session = session or requests.Session()
    domains = resolve_domains(brand)
    response = current_session.request(
        "GET",
        f"{domains.open_base}/open-apis/authen/v1/user_info",
        headers={"Authorization": f"Bearer {access_token}"},
        timeout=timeout,
    )
    if hasattr(response, "raise_for_status"):
        response.raise_for_status()
    payload = response.json()
    if isinstance(payload, dict) and payload.get("code", 0) not in (0, None):
        actual = (
            payload.get("data", {}).get("open_id")
            if isinstance(payload.get("data"), dict)
            else None
        )
        return IdentityVerificationResult(
            valid=False,
            expected_open_id=expected_open_id,
            actual_open_id=actual,
        )
    data = payload.get("data") or payload
    actual_open_id = data.get("open_id") if isinstance(data, dict) else None
    return IdentityVerificationResult(
        valid=actual_open_id == expected_open_id,
        expected_open_id=expected_open_id,
        actual_open_id=actual_open_id,
    )

File: src/orchestration.ts
import crypto from 'node:crypto';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { resolveDomains } from './domains.js';
import { DeviceAuthorization } from './models.js';
import {
  ContinuationState,
  FileContinuationStore,
  RuntimeCard,
  buildDeviceFlowCard,
  buildPermissionMissingCard,
} from './runtime-cards.js';
import { batchScopes, filterSensitiveScopes } from './scopes.js';

export type AuthErrorKind = 'app_scope_missing' | 'user_auth_required' | 'user_scope_insufficient';
export type AuthDecision = 'permission_card' | 'device_flow' | 'authorized';
export type ScopeNeedType = 'one' | 'all';
export type TokenType = 'tenant' | 'user';

function dedupePreserveOrder(items: string[]): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const item of items) {
    const normalized = item.trim();
    if (normalized && !seen.has(normalized)) {
      seen.add(normalized);
      result.push(normalized);
    }
  }
  return result;
}

export function defaultPendingFlowStorePath(): string {
  const configured = process.env.FEISHU_AUTH_KIT_PENDING_FLOW_STORE;
  if (configured) {
    return path.resolve(configured);
  }
  const stateHome = process.env.XDG_STATE_HOME;
  const base = stateHome ? path.resolve(stateHome) : path.join(os.homedir(), '.local', 'state');
  return path.join(base, 'feishu-auth-kit', 'pending_flows.json');
}

export interface PendingAuthFlowOptions {
  flowKey: string;
  operationId: string;
  appId: string;
  userOpenId?: string | null;
  decision: AuthDecision;
  requiredScopes?: string[];
  tokenType?: TokenType;
  scopeNeedType?: ScopeNeedType;
  status?: string;
  metadata?: Record<string, any>;
}

export class PendingAuthFlow {
  readonly flowKey: string;
  readonly operationId: string;
  readonly appId: string;
  readonly userOpenId: string | null;
  readonly decision: AuthDecision;
  readonly requiredScopes: string[];
  readonly tokenType: TokenType;
  readonly scopeNeedType: ScopeNeedType;
  readonly status: string;
  readonly metadata: Record<string, any>;

  constructor(options: PendingAuthFlowOptions) {
    this.flowKey = options.flowKey;
    this.operationId = options.operationId;
    this.appId = options.appId;
    this.userOpenId = options.userOpenId ?? null;
    this.decision = options.decision;
    this.requiredScopes = options.requiredScopes ? [...options.requiredScopes] : [];
    this.tokenType = options.tokenType ?? 'user';
    this.scopeNeedType = options.scopeNeedType ?? 'all';
    this.status = options.status ?? 'pending';
    this.metadata = options.metadata ? { ...options.metadata } : {};
  }

  toDict(): Record<string, any> {
    return {
      flow_key: this.flowKey,
      operation_id: this.operationId,
      app_id: this.appId,
      user_open_id: this.userOpenId,
      decision: this.decision,
      required_scopes: this.requiredScopes,
      token_type: this.tokenType,
      scope_need_type: this.scopeNeedType,
      status: this.status,
      metadata: this.metadata,
    };
  }
}

export interface PendingFlowUpsertResult {
  flow: PendingAuthFlow;
  reused: boolean;
  mergedScopes: string[];
}

export class FilePendingFlowRegistry {
  readonly path: string;

  constructor(filePath?: string | null) {
    this.path = filePath ? path.resolve(filePath) : defaultPendingFlowStorePath();
  }

  private _readAll(): Record<string, any> {
    if (!fs.existsSync(this.path)) {
      return {};
    }
    try {
      const text = fs.readFileSync(this.path, 'utf-8');
      const payload = JSON.parse(text);
      if (!payload || typeof payload !== 'object') {
        return {};
      }
      const items = payload.pending_flows || payload;
      if (!items || typeof items !== 'object') {
        return {};
      }
      const result: Record<string, any> = {};
      for (const [key, value] of Object.entries(items)) {
        if (value && typeof value === 'object') {
          result[key] = value;
        }
      }
      return result;
    } catch {
      return {};
    }
  }

  private _writeAll(items: Record<string, any>): void {
    const dir = path.dirname(this.path);
    fs.mkdirSync(dir, { recursive: true });
    const tempPath = `${this.path}.tmp.${Date.now()}`;
    fs.writeFileSync(
      tempPath,
      JSON.stringify({ pending_flows: items }, null, 2) + '\n',
      'utf-8',
    );
    fs.renameSync(tempPath, this.path);
  }

  load(flowKey: string): PendingAuthFlow | null {
    const item = this._readAll()[flowKey];
    if (!item) {
      return null;
    }
    return new PendingAuthFlow({
      flowKey: String(item.flow_key),
      operationId: String(item.operation_id),
      appId: String(item.app_id),
      userOpenId: item.user_open_id ?? null,
      decision: item.decision,
      requiredScopes: Array.isArray(item.required_scopes) ? item.required_scopes : [],
      tokenType: item.token_type ?? 'user',
      scopeNeedType: item.scope_need_type ?? 'all',
      status: String(item.status ?? 'pending'),
      metadata: item.metadata || {},
    });
  }

  remove(flowKey: string): boolean {
    const items = this._readAll();
    if (!(flowKey in items)) {
      return false;
    }
    delete items[flowKey];
    this._writeAll(items);
    return true;
  }

  upsert(flow: PendingAuthFlow): PendingFlowUpsertResult {
    const items = this._readAll();
    const existing = this.load(flow.flowKey);
    if (!existing) {
      items[flow.flowKey] = flow.toDict();
      this._writeAll(items);
      return {
        flow,
        reused: false,
        mergedScopes: [...flow.requiredScopes],
      };
    }

    const mergedScopes = dedupePreserveOrder([
      ...existing.requiredScopes,
      ...flow.requiredScopes,
    ]);
    const mergedFlow = new PendingAuthFlow({
      flowKey: existing.flowKey,
      operationId: existing.operationId,
      appId: existing.appId,
      userOpenId: flow.userOpenId || existing.userOpenId,
      decision: flow.decision,
      requiredScopes: mergedScopes,
      tokenType: flow.tokenType,
      scopeNeedType: flow.scopeNeedType,
      status: flow.status,
      metadata: { ...existing.metadata, ...flow.metadata },
    });
    items[flow.flowKey] = mergedFlow.toDict();
    this._writeAll(items);
    return {
      flow: mergedFlow,
      reused: true,
      mergedScopes,
    };
  }
}

export interface ScopeAuthorizationPlan {
  requestedScopes: string[];
  appGrantedScopes: string[];
  userGrantedScopes: string[];
  alreadyGrantedScopes: string[];
  missingUserScopes: string[];
  unavailableScopes: string[];
  batches: string[][];
}

export function planScopeAuthorization(options: {
  requestedScopes: string[];
  appGrantedScopes: string[];
  userGrantedScopes: string[];
  batchSize?: number;
  filterSensitive?: boolean;
}): ScopeAuthorizationPlan {
  const batchSize = options.batchSize ?? 100;
  const filterSensitive = options.filterSensitive ?? true;

  let requested = dedupePreserveOrder(options.requestedScopes);
  let appGranted = dedupePreserveOrder(options.appGrantedScopes);
  let userGranted = dedupePreserveOrder(options.userGrantedScopes);

  if (filterSensitive) {
    requested = filterSensitiveScopes(requested);
    appGranted = filterSensitiveScopes(appGranted);
    userGranted = filterSensitiveScopes(userGranted);
  }

  const appGrantedSet = new Set(appGranted);
  const userGrantedSet = new Set(userGranted);

  const unavailable = requested.filter((s) => !appGrantedSet.has(s));
  const available = requested.filter((s) => appGrantedSet.has(s));
  const already = available.filter((s) => userGrantedSet.has(s));
  const missing = available.filter((s) => !userGrantedSet.has(s));

  return {
    requestedScopes: requested,
    appGrantedScopes: appGranted,
    userGrantedScopes: userGranted,
    alreadyGrantedScopes: already,
    missingUserScopes: missing,
    unavailableScopes: unavailable,
    batches: batchScopes(missing, batchSize),
  };
}

export class AuthRequirement {
  readonly errorKind: AuthErrorKind;
  readonly requiredScopes: string[];
  readonly tokenType: TokenType;
  readonly scopeNeedType: ScopeNeedType;
  readonly userOpenId: string | null;
  readonly flowKey: string | null;
  readonly operationId: string | null;
  readonly metadata: Record<string, any>;

  constructor(options: {
    errorKind: AuthErrorKind;
    requiredScopes: string[];
    tokenType?: TokenType;
    scopeNeedType?: ScopeNeedType;
    userOpenId?: string | null;
    flowKey?: string | null;
    operationId?: string | null;
    metadata?: Record<string, any>;
  }) {
    this.errorKind = options.errorKind;
    this.requiredScopes = [...options.requiredScopes];
    this.tokenType = options.tokenType ?? 'user';
    this.scopeNeedType = options.scopeNeedType ?? 'all';
    this.userOpenId = options.userOpenId ?? null;
    this.flowKey = options.flowKey ?? null;
    this.operationId = options.operationId ?? null;
    this.metadata = options.metadata ? { ...options.metadata } : {};
  }
}

export class AuthContinuation {
  readonly operationId: string;
  readonly flowKey: string;
  readonly appId: string;
  readonly decision: AuthDecision;
  readonly userOpenId: string | null;
  readonly requiredScopes: string[];
  readonly tokenType: TokenType;
  readonly scopeNeedType: ScopeNeedType;
  readonly metadata: Record<string, any>;

  constructor(options: {
    operationId: string;
    flowKey: string;
    appId: string;
    decision: AuthDecision;
    userOpenId?: string | null;
    requiredScopes: string[];
    tokenType: TokenType;
    scopeNeedType: ScopeNeedType;
    metadata?: Record<string, any>;
  }) {
    this.operationId = options.operationId;
    this.flowKey = options.flowKey;
    this.appId = options.appId;
    this.decision = options.decision;
    this.userOpenId = options.userOpenId ?? null;
    this.requiredScopes = [...options.requiredScopes];
    this.tokenType = options.tokenType;
    this.scopeNeedType = options.scopeNeedType;
    this.metadata = options.metadata ? { ...options.metadata } : {};
  }

  toState(): ContinuationState {
    return {
      operationId: this.operationId,
      appId: this.appId,
      kind: this.decision,
      status: 'waiting',
      payload: {
        flow_key: this.flowKey,
        user_open_id: this.userOpenId,
        required_scopes: this.requiredScopes,
        token_type: this.tokenType,
        scope_need_type: this.scopeNeedType,
        metadata: this.metadata,
      },
    };
  }

  static fromState(state: ContinuationState): AuthContinuation {
    const payload = state.payload;
    return new AuthContinuation({
      operationId: state.operationId,
      flowKey: String(payload.flow_key),
      appId: state.appId,
      decision: state.kind as AuthDecision,
      userOpenId: payload.user_open_id ?? null,
      requiredScopes: Array.isArray(payload.required_scopes) ? payload.required_scopes : [],
      tokenType: payload.token_type ?? 'user',
      scopeNeedType: payload.scope_need_type ?? 'all',
      metadata: payload.metadata || {},
    });
  }
}

export function saveAuthContinuation(
  store: FileContinuationStore,
  continuation: AuthContinuation,
): ContinuationState {
  const state = continuation.toState();
  store.save(state);
  return state;
}

export function loadAuthContinuation(
  store: FileContinuationStore,
  operationId: string,
): AuthContinuation | null {
  const state = store.load(operationId);
  if (!state) {
    return null;
  }
  return AuthContinuation.fromState(state);
}

export interface RoutedAuthAction {
  decision: AuthDecision;
  flow: PendingAuthFlow;
  reusedExistingFlow: boolean;
  continuation: AuthContinuation;
  card: RuntimeCard | null;
  unavailableScopes: string[];
  batches: string[][];
}

export function routeAuthRequirement(options: {
  appId: string;
  requirement: AuthRequirement;
  pendingFlows: FilePendingFlowRegistry;
  continuationStore: FileContinuationStore;
  permissionUrl?: string | null;
  authorization?: DeviceAuthorization | null;
}): RoutedAuthAction {
  const { appId, requirement, pendingFlows, continuationStore, permissionUrl, authorization } = options;

  let decision: AuthDecision;
  if (requirement.errorKind === 'app_scope_missing') {
    decision = 'permission_card';
  } else {
    decision = 'device_flow';
  }

  const flowKey =
    requirement.flowKey ||
    `${appId}:${requirement.userOpenId || 'unknown'}:${decision}`;
  const operationId = requirement.operationId || crypto.randomUUID().replace(/-/g, '');

  const upsert = pendingFlows.upsert(
    new PendingAuthFlow({
      flowKey,
      operationId,
      appId,
      userOpenId: requirement.userOpenId,
      decision,
      requiredScopes: requirement.requiredScopes,
      tokenType: requirement.tokenType,
      scopeNeedType: requirement.scopeNeedType,
      metadata: requirement.metadata,
    }),
  );

  const continuation = new AuthContinuation({
    operationId: upsert.flow.operationId,
    flowKey: upsert.flow.flowKey,
    appId,
    decision,
    userOpenId: upsert.flow.userOpenId,
    requiredScopes: upsert.mergedScopes,
    tokenType: upsert.flow.tokenType,
    scopeNeedType: upsert.flow.scopeNeedType,
    metadata: upsert.flow.metadata,
  });
  saveAuthContinuation(continuationStore, continuation);

  if (decision === 'permission_card') {
    if (!permissionUrl) {
      throw new Error('permission_url is required for app_scope_missing routing');
    }
    const card = buildPermissionMissingCard({
      appId,
      operationId: upsert.flow.operationId,
      missingScopes: upsert.mergedScopes,
      permissionUrl,
      userOpenId: upsert.flow.userOpenId,
    });
    return {
      decision,
      flow: upsert.flow,
      reusedExistingFlow: upsert.reused,
      continuation,
      card,
      unavailableScopes: [],
      batches: [],
    };
  }

  if (!authorization) {
    throw new Error('authorization is required for user auth routing');
  }
  const card = buildDeviceFlowCard({
    appId,
    operationId: upsert.flow.operationId,
    authorization,
  });
  return {
    decision,
    flow: upsert.flow,
    reusedExistingFlow: upsert.reused,
    continuation,
    card,
    unavailableScopes: [],
    batches: [],
  };
}

export class SyntheticRetryArtifact {
  readonly schema: string;
  readonly kind: string;
  readonly operationId: string;
  readonly appId: string;
  readonly userOpenId: string | null;
  readonly text: string;
  readonly reason: string;
  readonly metadata: Record<string, any>;

  constructor(options: {
    operationId: string;
    appId: string;
    userOpenId?: string | null;
    text: string;
    reason: string;
    metadata?: Record<string, any>;
  }) {
    this.schema = 'feishu-auth-kit.synthetic-retry.v1';
    this.kind = 'synthetic_retry';
    this.operationId = options.operationId;
    this.appId = options.appId;
    this.userOpenId = options.userOpenId ?? null;
    this.text = options.text;
    this.reason = options.reason;
    this.metadata = options.metadata ? { ...options.metadata } : {};
  }

  toDict(): Record<string, any> {
    return {
      schema: this.schema,
      kind: this.kind,
      operation_id: this.operationId,
      app_id: this.appId,
      user_open_id: this.userOpenId,
      text: this.text,
      reason: this.reason,
      metadata: this.metadata,
    };
  }
}

export function buildSyntheticRetryArtifact(options: {
  operationId: string;
  appId: string;
  userOpenId?: string | null;
  text: string;
  reason: string;
  metadata?: Record<string, any>;
}): SyntheticRetryArtifact {
  return new SyntheticRetryArtifact(options);
}

export interface IdentityVerificationResult {
  valid: boolean;
  expectedOpenId: string;
  actualOpenId: string | null;
}

export async function verifyAccessTokenIdentity(options: {
  accessToken: string;
  expectedOpenId: string;
  brand?: string;
  timeout?: number;
}): Promise<IdentityVerificationResult> {
  const brand = options.brand ?? 'feishu';
  const timeoutMs = (options.timeout ?? 30) * 1000;
  const domains = resolveDomains(brand);

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const response = await fetch(`${domains.openBase}/open-apis/authen/v1/user_info`, {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${options.accessToken}`,
      },
      signal: controller.signal,
    });

    if (!response.ok) {
      throw new Error(`HTTP error ${response.status} ${response.statusText}`);
    }

    const payload = (await response.json()) as Record<string, any>;
    if (payload && payload.code !== 0 && payload.code !== undefined && payload.code !== null) {
      const actual = payload.data?.open_id ?? null;
      return {
        valid: false,
        expectedOpenId: options.expectedOpenId,
        actualOpenId: actual,
      };
    }

    const data = payload?.data || payload;
    const actualOpenId = data?.open_id ?? null;
    return {
      valid: actualOpenId === options.expectedOpenId,
      expectedOpenId: options.expectedOpenId,
      actualOpenId,
    };
  } finally {
    clearTimeout(timer);
  }
}
